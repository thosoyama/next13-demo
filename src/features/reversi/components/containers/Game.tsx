'use client'

import { DispatchContext, GameContext } from '../../context'
import { useGame } from '../../hooks/useGame'
import { Board } from '../parts/Board'
import { Info } from '../parts/Info'

/**
 * ゲーム
 */
export default function Game() {
  const [state, dispatch] = useGame()

  return (
    <>
      <GameContext.Provider value={state}>
        <DispatchContext.Provider value={dispatch}>
          <div className="m-auto flex aspect-square h-screen p-10">
            <Board />
            <Info />
          </div>
        </DispatchContext.Provider>
      </GameContext.Provider>
    </>
  )
}
