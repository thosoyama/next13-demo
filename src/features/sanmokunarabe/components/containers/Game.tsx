'use client'

import { useState } from 'react'

import { Board } from '../presentationals/Board'

type GameHistory = string[][]

export default function Game() {
  const [history, setHistory] = useState<GameHistory>([Array(9).fill('')])
  const [currentMove, setCurrentMove] = useState(0)
  const xIsNext = currentMove % 2 === 0
  const currentSquares = history[currentMove]

  function handlePlay(nextSquares: string[]) {
    const nextHistory = [...history.slice(0, currentMove + 1), nextSquares]
    setHistory(nextHistory)
    setCurrentMove(nextHistory.length - 1)
  }

  function jumpTo(nextMove: number) {
    setCurrentMove(nextMove)
  }

  const moves = history.map((_, move) => {
    let description
    if (move > 0) {
      description = `Go to move #${move}`
    } else {
      description = 'Go to game start'
    }
    return (
      // rome-ignore lint/suspicious/noArrayIndexKey: <explanation>
      <li key={move}>
        <button type="button" onClick={() => jumpTo(move)}>
          {description}
        </button>
      </li>
    )
  })

  return (
    <div className="game">
      <div className="game-board">
        <Board xIsNext={xIsNext} squares={currentSquares} onPlay={handlePlay} />
      </div>
      <div className="game-info">
        <ol>{moves}</ol>
      </div>
    </div>
  )
}
