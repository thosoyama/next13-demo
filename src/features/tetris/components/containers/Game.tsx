export function Game() {
  return <div>game</div>
}
