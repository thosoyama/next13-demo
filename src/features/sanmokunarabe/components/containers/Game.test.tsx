import '@testing-library/jest-dom'

import { fireEvent, render, screen } from '@testing-library/react'

import Game from './Game'

describe('Game', () => {
  it('renders a square', () => {
    render(<Game />)

    const squares: HTMLElement[] = screen.getAllByTestId('square')
    expect(squares.length).toBe(9)
    squares.forEach((square) => {
      expect(square).toBeInTheDocument()
      expect(square).toBeEmptyDOMElement()
    })
  })

  it('renders a square', () => {
    render(<Game />)

    const squares: HTMLElement[] = screen.getAllByTestId('square')

    fireEvent.click(squares[0])
    const actual1 = squares.map((square) => square.textContent)
    expect(actual1).toStrictEqual(['X', '', '', '', '', '', '', '', ''])

    fireEvent.click(squares[3])
    const actual2 = squares.map((square) => square.textContent)
    expect(actual2).toStrictEqual(['X', '', '', 'O', '', '', '', '', ''])

    fireEvent.click(squares[1])
    const actual3 = squares.map((square) => square.textContent)
    expect(actual3).toStrictEqual(['X', 'X', '', 'O', '', '', '', '', ''])

    fireEvent.click(squares[4])
    const actual4 = squares.map((square) => square.textContent)
    expect(actual4).toStrictEqual(['X', 'X', '', 'O', 'O', '', '', '', ''])

    fireEvent.click(squares[2])
    const actual5 = squares.map((square) => square.textContent)
    expect(actual5).toStrictEqual(['X', 'X', 'X', 'O', 'O', '', '', '', ''])
  })

  it('renders game unchanged', () => {
    const { container } = render(<Game />)
    expect(container).toMatchSnapshot()
  })
})
