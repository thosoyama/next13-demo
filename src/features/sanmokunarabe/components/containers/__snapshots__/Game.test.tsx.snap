// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders game unchanged 1`] = `
<div>
  <div
    class="game"
  >
    <div
      class="game-board"
    >
      <div
        class="status"
      >
        Next player: X
      </div>
      <div
        class="board-row"
      >
        <button
          class="square"
          data-testid="square"
          type="button"
        />
        <button
          class="square"
          data-testid="square"
          type="button"
        />
        <button
          class="square"
          data-testid="square"
          type="button"
        />
      </div>
      <div
        class="board-row"
      >
        <button
          class="square"
          data-testid="square"
          type="button"
        />
        <button
          class="square"
          data-testid="square"
          type="button"
        />
        <button
          class="square"
          data-testid="square"
          type="button"
        />
      </div>
      <div
        class="board-row"
      >
        <button
          class="square"
          data-testid="square"
          type="button"
        />
        <button
          class="square"
          data-testid="square"
          type="button"
        />
        <button
          class="square"
          data-testid="square"
          type="button"
        />
      </div>
    </div>
    <div
      class="game-info"
    >
      <ol>
        <li>
          <button
            type="button"
          >
            Go to game start
          </button>
        </li>
      </ol>
    </div>
  </div>
</div>
`;
